Firestore/Source/Local/FSTLocalDocumentsView.mm

```diff
@@ -110,37 +114,37 @@ - (FSTDocumentDictionary *)documentsMatchingDocumentQuery:(const ResourcePath &)
 }
 
 - (FSTDocumentDictionary *)documentsMatchingCollectionQuery:(FSTQuery *)query {
-  // Query the remote documents and overlay mutations.
-  // TODO(mikelehen): There may be significant overlap between the mutations affecting these
-  // remote documents and the allMutationBatchesAffectingQuery mutations. Consider optimizing.
+  // Get the remote documents in the state in which the backend last acknowledged them.
   __block FSTDocumentDictionary *results = [self.remoteDocumentCache documentsMatchingQuery:query];
-  results = [self localDocuments:results];
-
-  // Now use the mutation queue to discover any other documents that may match the query after
-  // applying mutations.
-  DocumentKeySet matchingKeys;
-  NSArray<FSTMutationBatch *> *matchingMutationBatches =
+  // Get locally persisted mutation batches.
+  NSArray<FSTMutationBatch *> *matchingBatches =
       [self.mutationQueue allMutationBatchesAffectingQuery:query];
-  for (FSTMutationBatch *batch in matchingMutationBatches) {
+
+  // Overlay mutations on top of the remote docs, and create docs that haven't yet been written to
+  // the backend.
+  for (FSTMutationBatch *batch in matchingBatches) {
     for (FSTMutation *mutation in batch.mutations) {
-      // TODO(mikelehen): PERF: Check if this mutation actually affects the query to reduce work.
+      // Only process documents belonging to the collection.
+      if (mutation.key.path().PopLast() != query.path) {
```

Comment:
PopLast() requires a full copy of the underlying array (less 1 segment). A better way to do this would be to add a method on Path that allows us to ask query.path if mutation.key.Path() is an immediate child.

Reply:
Done. The numbers I'm seeing in my local performance test are similar, but logically the dedicated method makes more sense.

Feel free to suggest changes on the naming. To quickly recap my reasoning: I think `IsChild` is slightly confusing because, to me at least, it's not clear whether it's the left or right side that is being checked, so that leaves `IsChildOf` vs `IsParentOf`. Since you mentioned asking `query.path`, I settled on `IsParentOf`.